src/Compression/CompressionCodecDeflate.cpp

```diff
@@ -0,0 +1,357 @@
+#include <cstdio>
+#include <thread>
+#include <Compression/CompressionCodecDeflate.h>
+#include <Compression/CompressionFactory.h>
+#include <Compression/CompressionInfo.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Poco/Logger.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int CANNOT_COMPRESS;
+    extern const int CANNOT_DECOMPRESS;
+}
+//DeflateJobHWPool
+qpl_job * DeflateJobHWPool::jobPool[jobPoolSize];
+std::atomic_bool DeflateJobHWPool::jobLock[jobPoolSize];
+
+DeflateJobHWPool & DeflateJobHWPool::instance()
+{
+    static DeflateJobHWPool ret;
+    return ret;
+}
+
+DeflateJobHWPool::DeflateJobHWPool()
+{
+    log = &Poco::Logger::get("DeflateJobHWPool");
+    if (initJobPool() < 0)
+    {
+        jobPoolEnabled = false;
+        LOG_WARNING(log, "DeflateJobHWPool is not ready. Please check if IAA hardware support.Auto switch to deflate software codec here");
+    }
+    else
+        jobPoolEnabled = true;
+}
+DeflateJobHWPool::~DeflateJobHWPool()
+{
+    destroyJobPool();
+}
+//HardwareCodecDeflate
+HardwareCodecDeflate::HardwareCodecDeflate()
+{
+    log = &Poco::Logger::get("HardwareCodecDeflate");
+    hwEnabled = DeflateJobHWPool::instance().jobPoolReady();
+}
+
+HardwareCodecDeflate::~HardwareCodecDeflate()
+{
+    if (!jobDecompAsyncMap.empty())
+    {
+        LOG_WARNING(log, "Find un-released job when HardwareCodecDeflate destroy");
+        for (auto it : jobDecompAsyncMap)
+        {
+            DeflateJobHWPool::instance().releaseJob(it.first);
+        }
+        jobDecompAsyncMap.clear();
+    }
+}
+uint32_t HardwareCodecDeflate::doCompressData(const char * source, uint32_t source_size, char * dest, uint32_t dest_size) const
+{
+    uint32_t job_id = 0;
+    qpl_job * job_ptr = DeflateJobHWPool::instance().acquireJob(&job_id);
+    if (job_ptr == nullptr)
+    {
+        LOG_WARNING(log, "HardwareCodecDeflate::doCompressData acquireJob fail!");
```

Comment:
We should throw here. Returning 0 means that we successfully compressed the input into 0 bytes.

Reply:
change this value to be static const: RET_ERROR which stands for hardware codec fail, need fallback to software codec.